File: pipeline/synthesizer.js
"use strict";

require("dotenv").config();

const PROMPT_TEMPLATE = `너는 교육 콘텐츠 분석 전문가야.

아래는 교육용 인터랙티브 콘텐츠에서 추출한 원본 텍스트야.
이 텍스트에는 HTML 태그, 코드, UI 요소(버튼, 메뉴, 네비게이션 텍스트) 등이
포함되어 있을 수 있어.

[지시사항]
1. HTML 태그, JavaScript, CSS 코드를 모두 무시해.
2. 버튼 라벨, 네비게이션 텍스트, UI 요소 텍스트를 무시해.
3. 교육적으로 의미 있는 내용만 파악해.
4. 해당 콘텐츠의 핵심 교육 내용을 한국어 3줄로 요약해.
5. 원본 텍스트에 없는 내용을 절대 추가하지 마.

규칙:
- 첫째 줄: 학습 주제 (무엇을 배우는가)
- 둘째 줄: 주요 학습 활동 (어떤 활동을 하는가)
- 셋째 줄: 학습 목표 및 기대 효과 (무엇을 할 수 있게 되는가)

[출력형식]
JSON 형식으로 응답할 것:
{ "summary": "line1. line2. line3" }

[원본 텍스트]
{CONTENT}`;

function buildPrompt(content, maxChars = 400000) {
  let text = content;
  if (text.length > maxChars) {
    text =
      text.slice(0, maxChars) +
      "\n\n[TRUNCATED — content exceeded context limit]";
  }
  return PROMPT_TEMPLATE.replace("{CONTENT}", text);
}

function parseResponse(raw) {
  try {
    return JSON.parse(raw);
  } catch {
    // Try extracting from markdown code block
    const match = raw.match(/```(?:json)?\s*([\s\S]*?)```/);
    if (match) {
      try {
        return JSON.parse(match[1].trim());
      } catch {
        /* fall through */
      }
    }
    // Try finding JSON object with "summary" key in text
    const objMatch = raw.match(
      /\{[\s\S]*?"summary"[\s\S]*?\}/,
    );
    if (objMatch) {
      try {
        return JSON.parse(objMatch[0]);
      } catch {
        /* fall through */
      }
    }
    return null;
  }
}

/** Call vLLM (OpenAI-compatible API) */
async function callVllm(prompt) {
  const baseUrl = process.env.VLLM_BASE_URL || "http://localhost:8000/v1";
  const model = process.env.VLLM_MODEL || "Qwen/Qwen3.5-35B-A3B";

  const res = await fetch(`${baseUrl}/chat/completions`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      model,
      messages: [{ role: "user", content: prompt }],
      temperature: 0,
      max_tokens: 512,
    }),
  });

  if (!res.ok) {
    const errText = await res.text();
    throw new Error(`vLLM error ${res.status}: ${errText}`);
  }

  const data = await res.json();
  return data.choices[0].message.content;
}

/** Call Gemini API via @google/genai */
async function callGemini(prompt) {
  const { GoogleGenAI } = require("@google/genai");
  const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY });
  const model = process.env.GEMINI_MODEL || "gemini-2.5-flash";

  const response = await ai.models.generateContent({
    model,
    contents: prompt,
    config: { temperature: 0, maxOutputTokens: 512 },
  });

  return response.text;
}

/** Main synthesis function */
async function synthesize(content) {
  const backend = process.env.BACKEND || "vllm";
  const prompt = buildPrompt(content);
  let raw;

  if (backend === "gemini") {
    raw = await callGemini(prompt);
  } else {
    raw = await callVllm(prompt);
  }

  const parsed = parseResponse(raw);
  if (!parsed) {
    return { error: "Failed to parse LLM response", raw };
  }

  return {
    summary: parsed.summary,
    raw,
  };
}

module.exports = {
  buildPrompt,
  parseResponse,
  callVllm,
  callGemini,
  synthesize,
};

File: pipeline/synthesizer.py
"""LLM synthesis via vLLM HTTP API.

Manages vLLM server lifecycle and sends pre-filtered content
to the model for summary extraction.
"""

import asyncio
import json
import os
import re
import signal
import subprocess
import sys
import time
from dataclasses import dataclass

import httpx

from config import (
    MAX_CONTENT_CHARS,
    MAX_TOKENS,
    SYSTEM_PROMPT,
    TEMPERATURE,
    VLLM_BASE_URL,
    VLLM_PORT,
    ModelConfig,
)


def build_messages(content: str) -> list[dict]:
    """Build chat messages for the LLM."""
    if len(content) > MAX_CONTENT_CHARS:
        content = content[:MAX_CONTENT_CHARS] + "\n\n[...truncated...]"

    return [
        {"role": "system", "content": SYSTEM_PROMPT},
        {"role": "user", "content": content},
    ]


def parse_llm_response(text: str) -> str:
    """Extract summary from LLM response."""
    text = re.sub(r'<think>[\s\S]*?</think>', '', text).strip()

    md_match = re.search(r'```(?:json)?\s*(\{[\s\S]*?\})\s*```', text)
    if md_match:
        text = md_match.group(1)

    try:
        json_match = re.search(r'\{[^{}]*"summary"[^{}]*\}', text)
        if json_match:
            data = json.loads(json_match.group())
            return data.get("summary", text)
        data = json.loads(text)
        return data.get("summary", text)
    except (json.JSONDecodeError, AttributeError):
        return text


async def call_llm(
    client: httpx.AsyncClient,
    model_id: str,
    content: str,
) -> dict:
    """Send content to vLLM and return parsed result."""
    messages = build_messages(content)
    payload = {
        "model": model_id,
        "messages": messages,
        "temperature": TEMPERATURE,
        "max_tokens": MAX_TOKENS,
    }

    start = time.monotonic()
    try:
        resp = await client.post(
            f"{VLLM_BASE_URL}/v1/chat/completions",
            json=payload,
            timeout=300.0,
        )
        latency_ms = int((time.monotonic() - start) * 1000)

        if resp.status_code != 200:
            return {"error": f"HTTP {resp.status_code}: {resp.text}", "latency_ms": latency_ms}

        data = resp.json()
        raw_text = data["choices"][0]["message"]["content"]
        usage = data.get("usage", {})

        return {
            "summary": parse_llm_response(raw_text),
            "raw_response": raw_text,
            "prompt_tokens": usage.get("prompt_tokens", 0),
            "completion_tokens": usage.get("completion_tokens", 0),
            "latency_ms": latency_ms,
        }
    except httpx.TimeoutException:
        return {"error": "Request timed out (300s)", "latency_ms": 300_000}
    except Exception as e:
        latency_ms = int((time.monotonic() - start) * 1000)
        return {"error": str(e), "latency_ms": latency_ms}


class VLLMServer:
    """Manages a vLLM server subprocess."""

    def __init__(self, model: ModelConfig, port: int = VLLM_PORT):
        self.model = model
        self.port = port
        self.process: subprocess.Popen | None = None

    def start(self) -> None:
        """Start the vLLM server and wait until it's healthy."""
        cmd = [
            sys.executable, "-m", "vllm.entrypoints.openai.api_server",
            "--model", self.model.model_id,
            "--port", str(self.port),
            *self.model.vllm_args,
        ]
        print(f"  Starting vLLM: {self.model.model_id}")
        print(f"  Command: {' '.join(cmd)}")

        self.process = subprocess.Popen(
            cmd,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            text=True,
        )

        if not self._wait_for_health(timeout=600):
            self.stop()
            raise RuntimeError(f"vLLM server failed to start for {self.model.model_id}")

        print(f"  vLLM server ready: {self.model.model_id}")

    def _wait_for_health(self, timeout: int = 600) -> bool:
        """Poll /v1/models until the server responds or timeout."""
        url = f"http://localhost:{self.port}/v1/models"
        deadline = time.monotonic() + timeout
        while time.monotonic() < deadline:
            try:
                resp = httpx.get(url, timeout=5.0)
                if resp.status_code == 200:
                    return True
            except httpx.ConnectError:
                pass
            if self.process and self.process.poll() is not None:
                stdout = self.process.stdout.read() if self.process.stdout else ""
                print(f"  vLLM process exited with code {self.process.returncode}")
                print(f"  Output: {stdout[-2000:]}")
                return False
            time.sleep(5)
        return False

    def stop(self) -> None:
        """Stop the vLLM server."""
        if self.process and self.process.poll() is None:
            print(f"  Stopping vLLM: {self.model.model_id}")
            self.process.terminate()
            try:
                self.process.wait(timeout=30)
            except subprocess.TimeoutExpired:
                self.process.kill()
                self.process.wait()
            print(f"  vLLM stopped.")
        self.process = None
